bug: flaky frontend unit suite — MissionObjectivesPanel debounce timer leaks across tests & fires after unmount

## Summary

The frontend `vitest` unit suite is **non-deterministically flaky** and intermittently turns the CI `frontend` check red. The same commit can pass then fail with no code change relevant to the failing test. This blocks the `frontend` required check on every PR (observed on PR #748, whose own diff only touched an unrelated e2e script).

## Symptom

CI failure (PR #748 re-run, run `26650841068`):

```
FAIL  __tests__/MissionObjectivesPanel.test.tsx > applies all 6 target fields when Apply is clicked (gh-601)
AssertionError: expected 18 to be close to 26, received difference is 8
 ❯ __tests__/MissionObjectivesPanel.test.tsx:133:43
   expect(lastCallArg.target_cruise_mps).toBeCloseTo(26.0, 3);
```

- `18` is the **default** `target_cruise_mps`; `26` is the **preset-applied** value.
- The test passes in isolation and passed in CI ~20 min earlier on identical relevant code.

## Root cause (confirmed + reproduced)

`MissionObjectivesPanel` schedules a **300 ms `setTimeout` debounce** on every `set()` call (including `mission_type` changes), stored in a `useRef` (`MissionObjectivesPanel.tsx:150-156`). **This timer is never cleared on unmount** — there is no `useEffect` cleanup.

Several tests (`shows the apply banner…`, `shows a diff row…`) trigger a `mission_type` change (scheduling the 300 ms timer) but never cancel it (no fake timers, no Apply click). `beforeEach` only does `updateMock.mockClear()` — it does **not** flush pending timers, and RTL's auto-unmount does not clear them because the component leaks them.

A leaked timer from an earlier test fires during the `await waitFor(...)` of the only async test (`applies all 6 target fields…`) and appends a **stale** `update({…trainerDraft, mission_type})` call (`target_cruise_mps: 18`) **after** the Apply call (`26`). The test reads `updateMock.mock.calls[last]` → `18` → fails. Pure timing → only manifests under CI load/scheduling.

### This is also a real production bug

The orphaned debounce fires **after the component unmounts**, issuing a stale-closure PUT for an aeroplane the user has already navigated away from. Deterministic reproduction (fake timers, on Node 22 = CI):

```tsx
const { unmount } = render(<MissionObjectivesPanel aeroplaneId="x" />);
fireEvent.change(missionTypeSelect, { target: { value: "motor_glider" } }); // schedules 300ms debounce
unmount();
vi.advanceTimersByTime(400);
expect(updateMock).not.toHaveBeenCalled();  // FAILS: update() called once with stale target_cruise_mps:18
```

## Proposed fix

Clear the pending debounce timer on unmount (cleanup-only `useEffect`, which does not violate the project's `react-hooks/set-state-in-effect` constraint):

```ts
useEffect(() => () => { if (timer.current) clearTimeout(timer.current); }, []);
```

This fixes the production bug (no stale write after unmount) **and** deterministically removes the cross-test timer leak (RTL's `afterEach` unmount now clears the timer), stabilising the CI flake.

## Severity

High — broken/unreliable test infra blocks the `frontend` required check on **every** frontend PR (Iron Law #6: a broken test environment is a blocker, not a footnote).

## Out of scope / related findings (separate follow-ups)

1. **Sibling uncleared-timer pattern** — `AssumptionRow.tsx`, `CreateWingDialog.tsx`, `hooks/usePreviewState.ts`, `hooks/useTessellation.ts` schedule `setTimeout`s without obvious unmount cleanup. Potential latent flakes; recommend an audit. Not fixed here (not observed flaking).
2. **Node-version brittleness** — the repo pins no Node version (`.nvmrc`/`engines` absent). On Node ≥ 24 (nvm default here is 24) the experimental global `localStorage` shadows jsdom's `window.localStorage`, making it `undefined` and crashing all localStorage-based tests (`ImportWarningBanner`, `StabilityOverlay`) locally. CI pins Node 22 so this does not affect CI, but it breaks local dev. Recommend adding `.nvmrc`=22 + `engines.node`.

## Acceptance criteria

- [ ] New test: `update()` is NOT called after `MissionObjectivesPanel` unmounts with a pending debounce (RED before fix, GREEN after).
- [ ] `MissionObjectivesPanel` clears its debounce timer on unmount.
- [ ] `npm run test:unit` passes deterministically across repeated full-suite runs.

--- a/frontend/__tests__/MissionObjectivesPanel.test.tsx
+++ b/frontend/__tests__/MissionObjectivesPanel.test.tsx
@@ -1,5 +1,5 @@
 import { describe, it, expect, vi, beforeEach } from "vitest";
-import { render, screen, fireEvent, waitFor } from "@testing-library/react";
+import { render, screen, fireEvent, waitFor, act } from "@testing-library/react";
 import React from "react";
 import { MissionObjectivesPanel } from "@/components/workbench/mission/MissionObjectivesPanel";
 
@@ -287,6 +287,32 @@ describe("MissionObjectivesPanel", () => {
         vi.useRealTimers();
       }
     });
+  });
+
+  // gh-763: the 300 ms debounce scheduled by set() (e.g. on a mission_type
+  // change) must be cleared when the component unmounts. Otherwise the
+  // orphaned timer fires after unmount and issues a stale-closure update()
+  // for an aeroplane the user has navigated away from — a data-write bug in
+  // production, and the source of cross-test timer leakage that flakes the
+  // full suite (the leaked call carries the pre-change draft and lands after
+  // a later test's assertions).
+  it("does not call update() after unmount when a debounce is pending (gh-763)", () => {
+    vi.useFakeTimers();
+    try {
+      const { unmount } = render(<MissionObjectivesPanel aeroplaneId="x" />);
+      const select = screen.getByLabelText(/Mission Type/i) as HTMLSelectElement;
+      fireEvent.change(select, { target: { value: "motor_glider" } });
+      // A 300 ms debounced update() is now pending. Unmount before it fires.
+      unmount();
+      // Advance well past the debounce window.
+      act(() => {
+        vi.advanceTimersByTime(400);
+      });
+      // The orphaned timer must have been cleared on unmount.
+      expect(updateMock).not.toHaveBeenCalled();
+    } finally {
+      vi.useRealTimers();
+    }
   });
 
   it("preserves in-flight draft edits across SWR revalidations for the SAME aeroplaneId (gh-602)", () => {

--- a/frontend/components/workbench/mission/MissionObjectivesPanel.tsx
+++ b/frontend/components/workbench/mission/MissionObjectivesPanel.tsx
@@ -1,6 +1,6 @@
 "use client";
 
-import React, { useRef, useState } from "react";
+import React, { useEffect, useRef, useState } from "react";
 import { useMissionObjectives, type MissionObjective } from "@/hooks/useMissionObjectives";
 import { useMissionPresets, type MissionPreset, type AxisName } from "@/hooks/useMissionPresets";
 import { normalizedToRaw } from "@/lib/missionScale";
@@ -130,6 +130,19 @@ export function MissionObjectivesPanel({ aeroplaneId }: Props) {
   const [bannerVisible, setBannerVisible] = useState<boolean>(false);
   const [lastAeroplaneId, setLastAeroplaneId] = useState<string | null>(null);
   const timer = useRef<ReturnType<typeof setTimeout> | null>(null);
+
+  // gh-763: clear any pending debounce on unmount. Without this, the 300 ms
+  // timer scheduled by set() fires after the component is gone and issues a
+  // stale-closure update() for an aeroplane the user already left — a spurious
+  // write in production, and the source of cross-test timer leakage that flakes
+  // the unit suite. Cleanup-only effect (no setState), so it does not run afoul
+  // of the react-hooks/set-state-in-effect constraint noted below.
+  useEffect(
+    () => () => {
+      if (timer.current) clearTimeout(timer.current);
+    },
+    [],
+  );
 
   // "Adjust state when a prop changes" pattern — avoids useEffect+setState
   // (which react-hooks/set-state-in-effect forbids). When the aeroplaneId

